### ISSUE TITLE
allow unused_mut ignored

### ISSUE THREAD

**Issue by C-Bouthoorn** [rust-lang/rust#40491]
```#[allow(unused_mut)]``` seems to be ignored

For example:
```
fn main() {
    #[allow(unused_mut)]
    let mut a = 10;

    println!("{}", a);
}
```

I expected to see no warning about the mut flag of a

Instead, this happened:

```
warning: variable does not need to be mutable, #[warn(unused_mut)] on by default
 --> TEST.rs:3:9
  |
3 |     let mut a = 10;
  |         ^^^^^
```

## Meta

This currently happens on all 3 major branches: stable, beta, nightly:

```
rustc 1.15.1 (021bd294c 2017-02-08)
binary: rustc
commit-hash: 021bd294c039bd54aa5c4aa85bcdffb0d24bc892
commit-date: 2017-02-08
host: x86_64-unknown-linux-gnu
release: 1.15.1
LLVM version: 3.9
```

```
rustc 1.16.0 (30cf806ef 2017-03-10)
binary: rustc
commit-hash: 30cf806ef8881c41821fbd43e5cf3699c5290c16
commit-date: 2017-03-10
host: x86_64-unknown-linux-gnu
release: 1.16.0
LLVM version: 3.9
```

```
rustc 1.17.0-nightly (824c9ebbd 2017-03-12)
binary: rustc
commit-hash: 824c9ebbd5f0611e326648d90ee3e50caf04bb7f
commit-date: 2017-03-12
host: x86_64-unknown-linux-gnu
release: 1.17.0-nightly
LLVM version: 3.9
```

Backtrace: `RUST_BACKTRACE=full` does not give more information.

**Comment by jmesmon** [comment]
Note that increasing the scope of the `allow` allows the error as expected: https://is.gd/RyBaRj

```
fn main() {
    #![allow(unused_mut)]
    let mut a = 10;

    println!("{}", a);
}
```

Unclear whether applying the `allow` to the `let` should be sufficient.

**Comment by sfackler** [comment]
I'm guessing that the lint passes were just not updated when attributes were allowed to be attached to statements.

**Comment by oli-obk** [comment]
The lint pass runs on functions, not on single statements, since you can't decide whether the `mut` should be linted just by looking at the statement. This is a failure of the current lint system and hard to address.

**Comment by oli-obk** [comment]
cc @Manishearth @mcarton @llogiq

**Comment by zetok** [comment]
>The lint pass runs on functions, not on single statements, since you can't decide whether the `mut` should be linted just by looking at the statement. This is a failure of the current lint system and hard to address.

`allow` is also ignored when `mut` statement is located inside of a macro, regardless of whether `allow` is located above the statement, macro, or function in which macro is called:
https://github.com/zetok/tox/blob/dc9107c599a33b2b851284808a73f02646ac948f/src/toxcore/dht_node.rs#L231,L235

It would be nice if it did work for single statements inside of a macro.

Main channels are still affected:
stable: rustc 1.20.0 (f3d6973f4 2017-08-27)
beta: rustc 1.21.0-beta.3 (ea1fd7d42 2017-09-15)
nightly: rustc 1.22.0-nightly (5dfc84cfa 2017-09-14)
